contrib/libxls/test/libxls_test.cc

```diff
@@ -0,0 +1,285 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "contrib/libxls/sandboxed.h"
+#include "contrib/libxls/utils/utils_libxls.h"
+#undef FILE  // XXX: artefect from genereted header
+
+#include "sandboxed_api/util/path.h"
+#include "sandboxed_api/util/status_matchers.h"
+
+namespace {
+
+using ::sapi::IsOk;
+using ::testing::Not;
+
+struct Sheet {
+  int last_row;
+  int last_col;
+  double values[4][4];
+};
+
+const struct TestVariant {
+  std::string filename;
+  size_t sheet_count;
+  struct Sheet sheet[2];
+} TestData[] = {
+    {
+        .filename = "t1.xls",
+        .sheet_count = 1,
+        .sheet =
+            {
+                {.last_row = 3,
+                 .last_col = 1,
+                 .values =
+                     {
+                         {1, 2, 0, 0},
+                         {3, 4, 0, 0},
+                         {5, 6, 0, 0},
+                         {7, 8, 0, 0},
+                     }},
+            },
+    },
+    {
+        .filename = "t2.xls",
+        .sheet_count = 2,
+        .sheet =
+            {
+                {.last_row = 1,
+                 .last_col = 2,
+                 .values =
+                     {
+                         {1, 2, 3, 0},
+                         {4, 5, 6, 0},
+                         {0, 0, 0, 0},
+                         {0, 0, 0, 0},
+                     }},
+                {.last_row = 1,
+                 .last_col = 1,
+                 .values =
+                     {
+                         {9, 8, 0, 0},
+                         {7, 6, 0, 0},
+                         {0, 0, 0, 0},
+                         {0, 0, 0, 0},
+                     }},
+            },
+    },
+};
+
+class LibXlsBase : public testing::Test {
+ protected:
+  std::string GetTestFilePath(const std::string& filename) {
+    return sapi::file::JoinPath(test_dir_, filename);
+  }
+
+  void SetUp() override;
+
+  const char* test_dir_;
+};
+
+class LibXlsTestFiles : public LibXlsBase,
+                        public testing::WithParamInterface<TestVariant> {};
+
+void LibXlsBase::SetUp() {
+  test_dir_ = getenv("TEST_FILES_DIR");
+  ASSERT_NE(test_dir_, nullptr);
+}
+
+TEST_P(LibXlsTestFiles, TestOpen) {
```

Comment:
most of the tests are just extension of the previous.
Should probably just leave the longest.

Reply:
I like the minimal to most complex approach.
But if you have strong opinion about this I can change that.

Reply:
I like the minimal to most complex approach.
But if you have strong opinion about this I can change that.

Reply:
If it's just pure extension it has little benefit (slightly better reporting), while it can increase test run time significantly